Unhelpful 'expected a type, got a overload group' error

When working on tensor code, several times I hit errors like these:

```
>       module = device.load_module(path)
E       slangpy.SlangCompileError: Failed to load slang module "test_tensor.slang"
E       (0): error 30060: expected a type, got a 'overload group'
E       (0): error 30060: expected a type, got a 'overload group'
E       (0): error 30060: expected a type, got a 'overload group'
E       (0): error 30060: expected a type, got a 'overload group'
E       (0): error 30060: expected a type, got a 'overload group'
E       (0): error 30060: expected a type, got a 'overload group'
E       (0): error 30060: expected a type, got a 'overload group'
E       (0): error 39999: import of module 'test_tensor.slang' failed because of a compilation error
E       (0): fatal error 39999: compilation ceased
E       abort compilation: (0): fatal error 39999: compilation ceased

slangpy\core\module.py:108: SlangCompileError
```

Which lead to divide/conquer until you find the (legitimate) compile error.

Reproduce branch in slangpy: dev/ccummings/tensor-refactor-overloadgroup
Run test_tensor.py:test_matrix_parameters

The actual error that it wasn't reporting correctly was a re-use of the same type alias twice. Once in tensor.slang, and another in primaltensor.slang:233.

--- a/source/slang/slang-check-type.cpp
+++ b/source/slang/slang-check-type.cpp
@@ -109,8 +109,25 @@ Type* SemanticsVisitor::getConstantBufferType(Type* elementType, Type* layoutTyp
     return m_astBuilder->getConstantBufferType(elementType, layoutType, witness);
 }
 
+static String _getExprName(Expr* expr)
+{
+    if (auto overloadedExpr = as<OverloadedExpr>(expr))
+    {
+        if (overloadedExpr->name)
+            return String(overloadedExpr->name->text);
+    }
+    if (auto declRefExpr = as<DeclRefExpr>(expr))
+    {
+        if (declRefExpr->name)
+            return String(declRefExpr->name->text);
+    }
+    return String();
+}
+
 Expr* SemanticsVisitor::ExpectATypeRepr(Expr* expr)
 {
+    auto originalExpr = expr;
+
     if (auto overloadedExpr = as<OverloadedExpr>(expr))
     {
         expr = resolveOverloadedExpr(overloadedExpr, LookupMask::type);
@@ -125,9 +142,17 @@ Expr* SemanticsVisitor::ExpectATypeRepr(Expr* expr)
         return expr;
     }
 
-    getSink()->diagnose(Diagnostics::ExpectedAType{
-        .whatWeGot = expr->type.type ? String(expr->type.type->toString()) : String("null"),
-        .expr = expr});
+    auto name = _getExprName(originalExpr);
+    if (name.getLength() == 0)
+        name = _getExprName(expr);
+
+    StringBuilder whatWeGot;
+    whatWeGot << "a '" << (expr->type.type ? expr->type.type->toString() : toSlice("null")) << "'";
+    if (name.getLength())
+        whatWeGot << " for '" << name << "'";
+
+    getSink()->diagnose(
+        Diagnostics::ExpectedAType{.whatWeGot = whatWeGot.produceString(), .expr = expr});
     return CreateErrorExpr(expr);
 }
 
@@ -284,8 +309,13 @@ bool SemanticsVisitor::CoerceToProperTypeImpl(
                 // diagnostic.
 
                 // Get the AST node type info, so we can output a 'got' name
+                auto exprName = _getExprName(originalExpr);
+                StringBuilder whatWeGotStr;
+                whatWeGotStr << "a '" << originalExpr->getClass().getName() << "'";
+                if (exprName.getLength())
+                    whatWeGotStr << " for '" << exprName << "'";
                 diagSink->diagnose(Diagnostics::ExpectedAType{
-                    .whatWeGot = originalExpr->getClass().getName(),
+                    .whatWeGot = whatWeGotStr.produceString(),
                     .expr = originalExpr});
             }
         }

--- a/source/slang/slang-diagnostics.lua
+++ b/source/slang/slang-diagnostics.lua
@@ -1297,7 +1297,7 @@ err(
     "expected-a-type",
     30060,
     "expected a type",
-    span { loc = "expr:Expr", message = "expected a type, got a '~whatWeGot'" }
+    span { loc = "expr:Expr", message = "expected a type, got ~whatWeGot" }
 )
 
 err(

--- a/tests/diagnostics/expected-type-overload-group.slang
+++ b/tests/diagnostics/expected-type-overload-group.slang
@@ -1,0 +1,12 @@
+//DIAGNOSTIC_TEST:SIMPLE(diag=CHECK): -target spirv
+
+// Regression test for https://github.com/shader-slang/slang/issues/9200
+// When an overload group is used where a type is expected, the diagnostic
+// should include the identifier name to help the user locate the problem.
+
+int MyFunc(int x) { return x; }
+float MyFunc(float x) { return x; }
+
+void test(MyFunc v) {}
+//CHECK:  ^^^^^^ expected a type
+//CHECK:  ^^^^^^ expected a type, got a 'overload group' for 'MyFunc'

